File: packages/core/src/__tests__/EventEmitter.test.ts
import { describe, it, expect, vi } from 'vitest';
import { EventEmitter } from '../event/EventEmitter.js';

describe('EventEmitter', () => {
  it('emits and receives events', () => {
    const emitter = new EventEmitter();
    const handler = vi.fn();

    emitter.on('test', handler);
    emitter.emit('test', { value: 1 });

    expect(handler).toHaveBeenCalledWith({ value: 1 });
  });

  it('supports multiple listeners', () => {
    const emitter = new EventEmitter();
    const h1 = vi.fn();
    const h2 = vi.fn();

    emitter.on('test', h1);
    emitter.on('test', h2);
    emitter.emit('test', { x: 1 });

    expect(h1).toHaveBeenCalledOnce();
    expect(h2).toHaveBeenCalledOnce();
  });

  it('removes a specific listener', () => {
    const emitter = new EventEmitter();
    const handler = vi.fn();

    emitter.on('test', handler);
    emitter.off('test', handler);
    emitter.emit('test', {});

    expect(handler).not.toHaveBeenCalled();
  });

  it('does not call removed listeners', () => {
    const emitter = new EventEmitter();
    const h1 = vi.fn();
    const h2 = vi.fn();

    emitter.on('test', h1);
    emitter.on('test', h2);
    emitter.off('test', h1);
    emitter.emit('test', {});

    expect(h1).not.toHaveBeenCalled();
    expect(h2).toHaveBeenCalledOnce();
  });

  it('does nothing when emitting with no listeners', () => {
    const emitter = new EventEmitter();
    expect(() => emitter.emit('nope', {})).not.toThrow();
  });

  it('removes all listeners for a specific event', () => {
    const emitter = new EventEmitter();
    const h1 = vi.fn();
    const h2 = vi.fn();

    emitter.on('a', h1);
    emitter.on('b', h2);
    emitter.removeAllListeners('a');
    emitter.emit('a', {});
    emitter.emit('b', {});

    expect(h1).not.toHaveBeenCalled();
    expect(h2).toHaveBeenCalledOnce();
  });

  it('removes all listeners', () => {
    const emitter = new EventEmitter();
    const h1 = vi.fn();
    const h2 = vi.fn();

    emitter.on('a', h1);
    emitter.on('b', h2);
    emitter.removeAllListeners();
    emitter.emit('a', {});
    emitter.emit('b', {});

    expect(h1).not.toHaveBeenCalled();
    expect(h2).not.toHaveBeenCalled();
  });

  it('reports listener count', () => {
    const emitter = new EventEmitter();
    expect(emitter.listenerCount('test')).toBe(0);

    const h = vi.fn();
    emitter.on('test', h);
    expect(emitter.listenerCount('test')).toBe(1);

    emitter.off('test', h);
    expect(emitter.listenerCount('test')).toBe(0);
  });
});

File: android/core/src/test/kotlin/com/devryner/rynbridge/core/EventEmitterTest.kt
package com.devryner.rynbridge.core

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class EventEmitterTest {

    @Test
    fun `on and emit`() {
        val emitter = EventEmitter()
        var received: Map<String, BridgeValue>? = null

        emitter.on("test") { data -> received = data }

        emitter.emit("test", mapOf("value" to BridgeValue.string("hello")))
        assertEquals("hello", received?.get("value")?.stringValue)
    }

    @Test
    fun `multiple listeners`() {
        val emitter = EventEmitter()
        var count = 0

        emitter.on("event") { _ -> count++ }
        emitter.on("event") { _ -> count++ }

        emitter.emit("event", emptyMap())
        assertEquals(2, count)
    }

    @Test
    fun `off removes listener`() {
        val emitter = EventEmitter()
        var count = 0

        val id = emitter.on("event") { _ -> count++ }
        emitter.emit("event", emptyMap())
        assertEquals(1, count)

        emitter.off("event", id)
        emitter.emit("event", emptyMap())
        assertEquals(1, count)
    }

    @Test
    fun `removeAllListeners for event`() {
        val emitter = EventEmitter()
        var count = 0

        emitter.on("event1") { _ -> count++ }
        emitter.on("event1") { _ -> count++ }
        emitter.on("event2") { _ -> count++ }

        emitter.removeAllListeners("event1")
        emitter.emit("event1", emptyMap())
        assertEquals(0, count)

        emitter.emit("event2", emptyMap())
        assertEquals(1, count)
    }

    @Test
    fun `removeAllListeners removes all`() {
        val emitter = EventEmitter()
        var count = 0

        emitter.on("event1") { _ -> count++ }
        emitter.on("event2") { _ -> count++ }

        emitter.removeAllListeners()
        emitter.emit("event1", emptyMap())
        emitter.emit("event2", emptyMap())
        assertEquals(0, count)
    }

    @Test
    fun `listenerCount`() {
        val emitter = EventEmitter()
        emitter.on("event") { _ -> }
        emitter.on("event") { _ -> }
        assertEquals(2, emitter.listenerCount("event"))
        assertEquals(0, emitter.listenerCount("other"))
    }

    @Test
    fun `emit nonexistent event does not crash`() {
        val emitter = EventEmitter()
        emitter.emit("nonexistent", emptyMap())
    }
}
